src/crypto/algorithms/megolm.ts

```diff
@@ -1419,154 +1438,318 @@ export class MegolmDecryption extends DecryptionAlgorithm {
+        const fromUs = event.getSender() === this.baseApis.getUserId();
+        const keyFromOurVerifiedDevice = deviceTrust.isVerified() && fromUs;
+        const weRequested = await this.wasRoomKeyRequested(event, roomKey);
+        const fromInviter = this.wasRoomKeyForwardedByInviter(event, roomKey);
+        const sharedAsHistory = this.wasRoomKeyForwardedAsHistory(roomKey);
+
+        return (weRequested && keyFromOurVerifiedDevice) || (fromInviter && sharedAsHistory);
+    }
+
+    /**
+     * Did we ever request the given room key from the event sender and its
+     * accompanying device.
+     *
+     * @param event - An `m.forwarded_room_key` event.
+     * @param roomKey - The room key that was found in the event.
+     *
+     * @internal
+     *
+     */
+    private async wasRoomKeyRequested(event: MatrixEvent, roomKey: RoomKey): Promise<boolean> {
+        // We send the `m.room_key_request` out as a wildcard to-device request,
+        // otherwise we would have to duplicate the same content for each
+        // device. This is why we need to pass in "*" as the device id here.
+        const outgoingRequests = await this.crypto.cryptoStore.getOutgoingRoomKeyRequestsByTarget(
+            event.getSender()!,
+            "*",
+            [RoomKeyRequestState.Sent],
+        );
+
+        return outgoingRequests.some(
+            (req) => req.requestBody.room_id === roomKey.roomId && req.requestBody.session_id === roomKey.sessionId,
+        );
+    }
+
+    private wasRoomKeyForwardedByInviter(event: MatrixEvent, roomKey: RoomKey): boolean {
+        // TODO: This is supposed to have a time limit. We should only accept
+        // such keys if we happen to receive them for a recently joined room.
+        const room = this.baseApis.getRoom(roomKey.roomId);
+        const senderKey = event.getSenderKey();
+
+        if (!senderKey) {
+            return false;
+        }
+
+        const senderKeyUser = this.crypto.deviceList.getUserByIdentityKey(olmlib.OLM_ALGORITHM, senderKey);
+
+        if (!senderKeyUser) {
+            return false;
+        }
+
+        const memberEvent = room?.getMember(this.userId)?.events.member;
+        const fromInviter =
+            memberEvent?.getSender() === senderKeyUser ||
+            (memberEvent?.getUnsigned()?.prev_sender === senderKeyUser &&
+                memberEvent?.getPrevContent()?.membership === "invite");
+
+        if (room && fromInviter) {
+            return true;
         } else {
-            keysClaimed = event.getKeysClaimed();
+            return false;
         }
+    }
 
-        if (content["org.matrix.msc3061.shared_history"]) {
-            extraSessionData.sharedHistory = true;
+    private wasRoomKeyForwardedAsHistory(roomKey: RoomKey): boolean {
+        const room = this.baseApis.getRoom(roomKey.roomId);
+
+        if (room && roomKey.extraSessionData.sharedHistory) {
+            return true;
+        } else {
+            return false;
+        }
```

Comment:
making this dependent on `room` feels surprising, given its name. I'm also a bit unclear why it is needed: `wasRoomKeyForwardedByInviter` will return false for an unknown room, so `(fromInviter && sharedAsHistory)` already checks for a known room via `fromInviter`. Why not set `sharedAsHistory = roomKey.extraSessionData.sharedHistory` ?

Reply:
I'm not sure that it needs to check for the room either but I don't think that because some other method, that coincidentally gets called alongside this one, checks for a room is a particularly good reason to remove the check.

I think what we should be asking us is the following: "Do we want to consider forwarded room keys shared as history if we're not part of the room".

This flag is controlled by the user that does the forwarding, in the spirit of minimizing the amount of forwarded keys we accept, I decided that we only consider forwarded room keys to be accepted as historic ones only if there's a room present.

No room, no room history and the flag is meaningless to us. It makes it a bit harder for users to try to insert arbitrary room keys into our store. 

Now the point is mostly moot, since the homeserver controls our room membership and it can insert us into a room to get us to consider the flag anyways. This check would make a bit more sense in a rosy world where the homeserver can't insert us  into arbitrary rooms.

Considering the above I'm fine with removing it, though I'll remove the whole method in that case and just access the flag directly in the `(fromInviter && sharedAsHistory)` line. That way the method can't be used in a place where there's no `room` check alongside it.

Reply:
> I'll remove the whole method in that case and just access the flag directly in the `(fromInviter && sharedAsHistory)` line.

That sounds like a good approach to me.

I don't have a strong feeling here, but if you want to keep the method, I would ask that you add a comment explaining some of this:

```javascript
// If the key is not for a known room, then something fishy is going on, so we reject the key out of caution.
// In practice, this is a bit moot because we'll only accept shared_history forwarded by the inviter, and we won't 
// know who was the inviter for an unknown room, so we'll reject it anyway.
if (!room) {
   return false;
}
```

Reply:
I opted for the additional comment in the end https://github.com/matrix-org/matrix-js-sdk/pull/2982/commits/dd4973fd4c0ede7b195f3a7a584f376d925d52a9.